drivers/bus/acpi/cmbatt/cmbatt.c

```diff
@@ -292,16 +294,73 @@ CmBattVerifyStaticInfo(PCMBATT_DEVICE_EXTENSION DeviceExtension,
         // FIXME: take from _BIX method: Info->CycleCount
         DeviceExtension->BifData = BifData;
 
-        if (BifData.PowerUnit == 1)
+        /* Check if the power stats are reported in ampere or watts */
+        if (BifData.PowerUnit == ACPI_BATT_POWER_UNIT_AMPS)
         {
-            DPRINT1("FIXME: need to convert mAh into mWh\n");
-            Info->DesignedCapacity = BATTERY_UNKNOWN_CAPACITY;
-            Info->FullChargedCapacity = BATTERY_UNKNOWN_CAPACITY;
-            Info->DefaultAlert1 = BATTERY_UNKNOWN_CAPACITY;
-            Info->DefaultAlert2 = BATTERY_UNKNOWN_CAPACITY;
+            /*
+             * We have got power stats in milli-ampere but ReactOS expects the values
+             * to be reported in milli-watts, so we have to convert them.
+             * In order to do so we must expect the design voltage of the battery
+             * is not unknown.
+             */
+            DesignVoltage = DeviceExtension->BifData.DesignVoltage;
+            if ((DesignVoltage != BATTERY_UNKNOWN_VOLTAGE) && (DesignVoltage))
+            {
+                /* Convert the design capacity */
+                if (BifData.DesignCapacity != BATTERY_UNKNOWN_CAPACITY)
+                {
+                    Info->DesignedCapacity = CONVERT_MAH_TO_MWH(BifData.DesignCapacity, DesignVoltage);
+                }
+                else
+                {
+                    Info->DesignedCapacity = BATTERY_UNKNOWN_CAPACITY;
+                }
+
+                /* Convert the full charged capacity */
+                if (BifData.LastFullCapacity != BATTERY_UNKNOWN_CAPACITY)
+                {
+                    Info->FullChargedCapacity = CONVERT_MAH_TO_MWH(BifData.LastFullCapacity, DesignVoltage);
+                }
+                else
+                {
+                    Info->FullChargedCapacity = BATTERY_UNKNOWN_CAPACITY;
+                }
+
+                /* Convert the low capacity alarm (DefaultAlert1) */
+                if (BifData.DesignCapacityLow != BATTERY_UNKNOWN_CAPACITY)
+                {
+                    Info->DefaultAlert1 = CONVERT_MAH_TO_MWH(BifData.DesignCapacityLow, DesignVoltage);
+                }
+                else
+                {
+                    Info->DefaultAlert1 = BATTERY_UNKNOWN_CAPACITY;
+                }
+
+                /* Convert the designed capacity warning alarm (DefaultAlert2) */
+                if (BifData.DesignCapacityWarning != BATTERY_UNKNOWN_CAPACITY)
+                {
+                    Info->DefaultAlert2 = CONVERT_MAH_TO_MWH(BifData.DesignCapacityWarning, DesignVoltage);
+                }
+                else
+                {
+                    Info->DefaultAlert2 = BATTERY_UNKNOWN_CAPACITY;
+                }
```

Comment:
These four conditional assignments are so similar and screaming like... MAKE ME A FUNCTION 🙀

Reply:
No.

Reply:
Or a macro
```c
#define CONVERT_BATT_INFO(BifData, DesignVoltage) \
    ((BifData) != BATTERY_UNKNOWN_CAPACITY) \
        ? CONVERT_MAH_TO_MWH((BifData), (DesignVoltage)) \
        : BATTERY_UNKNOWN_CAPACITY
```
to be used for example, as:
```c
Info->DefaultAlert2 = CONVERT_BATT_INFO(BifData.DesignCapacityWarning, DesignVoltage);
```

Note that this "CONVERT_BATT_INFO" macro could actually be merged together with the "CONVERT_MAH_TO_MWH" one.

Reply:
Any particular reason not to do that? Same could be achieved with a define:
```c
                #define CONVERT_CAPACITY(x) (BifData.(x) == BATTERY_UNKNOWN_CAPACITY ? BATTERY_UNKNOWN_CAPACITY : CONVERT_MAH_TO_MWH(BifData.(x), DesignVoltage))

                /* Convert the design capacity */
                Info->DesignedCapacity = CONVERT_CAPACITY(DesignCapacity);

                /* Convert the full charged capacity */
                Info->FullChargedCapacity = CONVERT_CAPACITY(LastFullCapacity);

                /* Convert the low capacity alarm */
                Info->DefaultAlert1 = CONVERT_CAPACITY(DesignCapacityLow);

                /* Convert the designed capacity warning alarm (DefaultAlert2) */
                Info->DefaultAlert2 = CONVERT_CAPACITY(DesignCapacityWarning);
```

Reply:
It's better now, thanks!